sse2neon.h

```diff
@@ -1405,38 +1763,34 @@ FORCE_INLINE __m128i _mm_srai_epi16(__m128i a, int count)
         ret;                                                            \
     })
 
-// NEON does not provide a version of this function, here is an article about
-// some ways to repro the results.
-// http://stackoverflow.com/questions/11870910/sse-mm-movemask-epi8-equivalent-method-for-arm-neon
+
+// NEON does not provide a version of this function.
 // Creates a 16-bit mask from the most significant bits of the 16 signed or
 // unsigned 8-bit integers in a and zero extends the upper bits.
 // https://msdn.microsoft.com/en-us/library/vstudio/s090c8fk(v=vs.100).aspx
 FORCE_INLINE int _mm_movemask_epi8(__m128i _a)
 {
+    // Use increasingly wide shifts+adds to collect the sign bits
+    // together.
+    // 89 FF 1D C0 00 10 99 33
     uint8x16_t input = vreinterpretq_u8_m128i(_a);
-    static const int8_t __attribute__((aligned(16)))
-    xr[8] = {-7, -6, -5, -4, -3, -2, -1, 0};
-    uint8x8_t mask_and = vdup_n_u8(0x80);
-    int8x8_t mask_shift = vld1_s8(xr);
-
-    uint8x8_t lo = vget_low_u8(input);
-    uint8x8_t hi = vget_high_u8(input);
-
-    lo = vand_u8(lo, mask_and);
-    lo = vshl_u8(lo, mask_shift);
-
-    hi = vand_u8(hi, mask_and);
-    hi = vshl_u8(hi, mask_shift);
-
-    lo = vpadd_u8(lo, lo);
-    lo = vpadd_u8(lo, lo);
-    lo = vpadd_u8(lo, lo);
-
-    hi = vpadd_u8(hi, hi);
-    hi = vpadd_u8(hi, hi);
-    hi = vpadd_u8(hi, hi);
-
-    return ((hi[0] << 8) | (lo[0] & 0xFF));
+    // Shift out everything but the sign bits
+    // 01 01 00 01 00 00 01 00
+    uint16x8_t high_bits = vreinterpretq_u16_u8(vshrq_n_u8(input, 7));
+    // Merge the even lanes together. The 'xx' bits are garbage.
+    // xx03 xx02 xx00 xx01
```

Comment:
I thought it should be `xx03 xx01 xx00 xx02`?

Reply:
You wanted them not to be printed in little endian order, so I reversed them. Now you are more confused.

Maybe I should use this diagram:
```
89 ff 1d c0 00 10 99 33
 \  \  \  \  \  \  \  \
01 01 00 01 00 00 01 00
 |_/   |_/   |_/   |_/
/|    /|    /|    /|
\|    \|    \|    \|
03 01 02 01 00 00 01 00
03 xx 02 xx 00 xx 01 xx
 |____/      |_____/
/|          /|
\|          \|
0b 05 02 01 04 00 01 00
0b xx xx xx 04 xx xx xx
 |__________/
/|
||
4b 05 12 01 04 00 01 00
4b xx xx xx xx xx xx xx
4b
```
This shows the actual byte values of the example vector and how the shifts work in little endian. However, while it would be backwards, the algorithm is easy to see in big endian:
```
89 ff 1d c0 00 10 99 33
 \  \  \  \  \  \  \  \
01 01 00 01 00 00 01 00
  \ |   \ |   \ |   \ |
   \|    \|    \|    \|
xx 03 xx 01 xx 00 xx 02
     \___ |     \____ |
         \|      \   \|
xx xx xx 0d xx xx xx 02
           \_________ |
                     ||
xx xx xx xx xx xx xx d2
d2
```

I should use big endian as it is clear, but make a note that it is actually done in little endian.